crates/ruma-client-api/src/r0/keys/upload_signatures.rs

```diff
@@ -19,25 +24,76 @@ ruma_api! {
     request: {
         /// Signed keys.
         #[ruma_api(body)]
-        pub signed_keys: BTreeMap<Box<UserId>, BTreeMap<String, JsonValue>>,
+        pub signed_keys: BTreeMap<Box<UserId>, SignedKeys>,
     }
 
     #[derive(Default)]
-    response: {}
+    response: {
+        /// Signature processing failures.
+        pub failures: BTreeMap<Box<UserId>, BTreeMap<String, Failure>>,
+    }
 
     error: crate::Error
 }
 
 impl Request {
     /// Creates a new `Request` with the given signed keys.
-    pub fn new(signed_keys: BTreeMap<Box<UserId>, BTreeMap<String, JsonValue>>) -> Self {
+    pub fn new(signed_keys: BTreeMap<Box<UserId>, SignedKeys>) -> Self {
         Self { signed_keys }
     }
 }
 
 impl Response {
     /// Creates an empty `Response`.
     pub fn new() -> Self {
-        Self {}
+        Self::default()
+    }
+}
+
+/// A map of key IDs to signed key objects.
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+#[serde(transparent)]
+pub struct SignedKeys(BTreeMap<Box<str>, Box<RawJsonValue>>);
+
+impl SignedKeys {
+    /// Creates an empty `SignedKeys` map.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Add the given device keys.
+    pub fn add_device_keys(&mut self, device_id: Box<DeviceId>, device_keys: Raw<DeviceKeys>) {
+        self.0.insert(device_id.into(), device_keys.into_json());
     }
+
+    /// Add the given cross signing keys.
+    pub fn add_cross_signing_keys(
+        &mut self,
+        what_the_hell_is_this: Box<str>,
```

Comment:
So this is my main problem.. What is this key? The example has two of these, `base64+master+public+key` and `bobs+base64+master+public+key`.

Reply:
It's again the base64 encoded public key of the individual cross signing key. I think we can call this a key ID.

Reply:
I want to call it the public key if that's what it is. We already have a bunch of key IDs in Ruma and those are pairs of an algorithm and an identifier separated by `:`. (see https://docs.ruma.io/ruma_identifiers/index.html#types)

Reply:
Well, semantically, it is the thing that is used to reference the public key. Incidentally this is the same thing as the public key but it's the `DeviceId` equivalent for a single cross signing key.